code/cfile/cfilesystem.cpp

```diff
@@ -1753,42 +2021,36 @@ int cf_get_file_list(int max, char** list, int pathtype, const char* filter, int
 				}
 			}
 
-			if (num_files >= max)
-				break;
-			
-			if ( !cf_matches_spec(filter, f->name_ext.c_str())) {
+			if ( !glob && !sub_path_match(sub_path, f->sub_path) ) {
 				continue;
 			}
 
-			if ( cf_file_already_in_list(num_files, list, f->name_ext.c_str()) ) {
+			if ( !cf_matches_spec(filter.c_str(), f->name_ext.c_str())) {
 				continue;
 			}
 
-			if (Skip_packfile_search && f->pack_offset != 0) {
-				// If the packfile skip flag is set we skip files in VPs but still search in directories
-				continue;
+			auto pos = f->name_ext.rfind('.');
+
+			if ( !sub_path.empty() ) {
+				// prepend file name with sub directory path
+				// allows us to open the specific file and allows duplicate filenames in different paths
+				fullname = f->sub_path + f->name_ext.substr(0, pos);
+			} else {
+				fullname = f->name_ext.substr(0, pos);
 			}
 
-			if (Skip_memory_files && f->data != nullptr) {
-				// If we want to skip memory files and this is a memory file then ignore it
+			if ( cf_file_already_in_list(num_files, list, fullname.c_str()) ) {
 				continue;
 			}
 
 			if ( !Get_file_list_filter || (*Get_file_list_filter)(f->name_ext.c_str()) ) {
-
 				//mprintf(( "Found '%s' in root %d path %d\n", f->name_ext, f->root_index, f->pathtype_index ));
 
-					auto pos = f->name_ext.rfind('.');
-
-					if (pos != SCP_string::npos) {
-						l = pos;
-					} else {
-						l = f->name_ext.length();
-					}
+				auto len = fullname.length();
 
-					list[num_files] = (char *)vm_malloc(l + 1);
-					strncpy(list[num_files], f->name_ext.c_str(), l);
-					list[num_files][l] = 0;
+				list[num_files] = (char *)vm_malloc(len + 1);
+				strncpy(list[num_files], f->name_ext.c_str(), len);
+				list[num_files][len] = 0;
```

Comment:
Why are you allocating a string of the length of fullname and then copying name_ext into it?
Should this whole block not just be ``list[num_files] = vm_strdup(f->name_ext.c_str());``?

Reply:
That was how Volition did it and I didn't even think of changing it while working on this. Good time to get that fixed though!

Reply:
This actually pointed out a bug too since it should have been using the generated `fullname` rather than `f->name_ext` here. That would have been a pretty annoying bug to track down if it had slipped through.